src/parseTree.js

```diff
@@ -281,37 +270,121 @@
+      // from currentLR
+      currentLR: {type: Object, default: Object},
+      eventHandlers: {type: Object}
     },
     computed: {
+      labeled: function() {
+        return shouldNodeBeLabeled(this.traceNode);
+      },
+      vbox: function() {
+        return hasVisibleChoice(this.traceNode) ||
+               hasVisibleLeftRecursion(this.traceNode) ||
+               (isAlt(this.traceNode.expr) && this.context && this.context.vbox);
+      },
+      isWhitespace: function() {
+        return this.traceNode.ruleName === 'spaces';
+      },
+      isLeaf: function() {
+        var leaf = isLeaf(this.traceNode);
+        if (this.traceNode.isMemoized) {
+          var memoKey = this.traceNode.expr.toMemoKey();
+          var stack = this.currentLR[memoKey];
+          if (stack && stack[stack.length - 1] === this.traceNode.pos) {
+            leaf = true;
+          }
+        }
+        return leaf;
+      },
       classObj: function() {
         var obj = {
-          disclosure: this.classes.labeled && this.isInVBox
+          disclosure: this.labeled && this.isInVBox
         };
         var ctorName = this.traceNode.ctorName;
         if (ctorName) {
           obj[ctorName.toLowerCase()] = true;
         }
-        Object.assign(obj, this.classes);
+
+        obj.collapsed = this.labeled &&
+          this.context &&
+          shouldTraceElementBeCollapsed(this.traceNode, this.context);
+        obj.failed = !this.traceNode.succeeded;
+        obj.labeled = this.labeled;
+        obj.leaf = this.isLeaf;
+        // Object.assign(obj, this.classes);
         return obj;
       },
+      children: function() {
+        this.$data.childrenNodes = this.classObj.collapsed ? undefined : [];
+        var self = this;
+        function enter(node, parent, depth) {
+          // Don't show or recurse into nodes that failed, unless "Show failures" is enabled.
+          if ((!node.succeeded && !ohmEditor.options.showFailures) ||
+              (node.isImplicitSpaces && !ohmEditor.options.showSpaces)) {
+            return node.SKIP;
+          }
+          // Don't bother showing whitespace nodes that didn't consume anything.
+          var isWhitespace = node.expr.ruleName === 'spaces';
+          if (isWhitespace && node.source.contents.length === 0) {
+            return node.SKIP;
+          }
+
+          var traceElement = {
+            traceNode: node,
+            context: {
+              parent: self,
+              collapsed: self.collapsed,
+              syntactic: self.labeled ? isSyntactic(node.expr) :
+                                        self.context && self.context.syntactic,
+              vbox: self.vbox
+            },
+            isInVBox: self.context ? self.context.vbox : false,
+            currentLR: self.currentLR
+          };
+          self.$data.childrenNodes.push(traceElement);
+          return node.SKIP;
+        }
+
+        if (this.$data.childrenNodes) {
+          if (this.traceNode.isRootNode) {
```

Comment:
In fact, since we're no longer doing a recursive walk, we don't need to call `walk` at all! I'd rename `enter` to `visitTraceNode`, and then you can just visit the children like this:

```
var children = [];
function visitTraceNode(node) {
  ...
  children.push(traceElement);
}
this.traceNode.children.forEach(visitTraceNode);
```

Then you can also drop the `parent` and `depth` arguments, and you don't need `node.SKIP` anymore. I think that will make this code a lot simpler.